packages/video_player/video_player/android/src/main/java/io/flutter/plugins/videoplayer/VideoPlayer.java

```diff
@@ -279,6 +279,9 @@ private void sendInitialized() {
           width = exoPlayer.getVideoFormat().height;
           height = exoPlayer.getVideoFormat().width;
         }
+        if (rotationDegrees == 180) {
+          event.put("rotation", Math.PI);
+        }
```

Comment:
This seems ok IMHO, but I'd prefer to rotate the video correctly within the Java code instead of rotating the flutter VideoPlayer widget.

One option I started looking into is using a [TextureView](https://source.android.com/devices/graphics/arch-tv) with [exoPlayer.setVideoTextureView](https://google.github.io/ExoPlayer/doc/reference/com/google/android/exoplayer2/SimpleExoPlayer.html#setVideoTextureView-android.view.TextureView-) instead of a [Surface](https://developer.android.com/reference/android/view/Surface.html) on line 170 in this file. (The top answers on these old StackOverflow questions [[1](https://stackoverflow.com/questions/48841506/rotating-exoplayer), [2](https://stackoverflow.com/questions/44564356/how-to-rotate-exoplayer)] about rotating exoplayer and [the docs](https://source.android.com/devices/graphics/arch-tv#surface_or_texture) comparing SurfaceView and TextureView suggest TextureView is better, or required, for rotation. The docs suggest TextureView can reduce performance because of duplicate compositing and shouldn't be used in newer APIs.)

It seems switching from Surface to TextureView would be a much larger refactoring of how the plugin works. Combined with the performance concerns of TextureView, maybe the Flutter rotation isn't such a bad solution.

I do want to look more into why these videos aren't upside down in the ExoPlayer sample app (according to cyanglaz's [comment](https://github.com/flutter/flutter/issues/60327#issuecomment-713019545)). Does it use a TextureView?

Reply:
I think it's reasonable to move forward with this, but let's add a comment here since it's non-obvious outside the context of this PR why this is only being sent for one specific rotation. E.g.:

```
// Rotating the video with ExoPlayer does not seem to be possible with a Surface,
// so inform the Flutter code that the widget needs to be rotated to prevent
// upside-down playback.
```

Reply:
Although actually: what is the behavior here for the 90/270 cases? Does the video play sideways, or is that handled differently by ExoPlayer than 180?

I'm wondering if we should always send the rotation, so that more information is available to plugin clients on the Dart side.

Reply:
I think I'll just rename `rotation` to be `rotationCorrection` and add a comment as you suggested.

I confirmed videos recorded in 0, 90, and 270 orientations are displayed correctly without this PR (only 180 needs to be fixed).

I like the idea of passing along more information to clients, but it seems a bit messy to add a `rotation` value that's currently only populated on Android and only has an effect in the case of 180 on Android.
(If we do pass it along, the `Transform.rotate` widget in video_player.dart would potentially have:
`angle: widget.controller.value.rotation == 180 ? pi : 0.0,`)

Another argument is that clients shouldn't have to care about the recording orientation if they know the default display orientation is correct and they can compare height/width if they want to treat portrait/landscape differently.

WDYT?

Reply:
(Done in https://github.com/flutter/plugins/pull/3820/commits/5f95287a54b82ca90cde6153285ab9e9b10e2493)